### ISSUE TITLE
Script attribute value with quotes is escaped

### ISSUE THREAD

**Issue by aSemy** [facebook/docusaurus#10172]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

I want to use [Cloudflare Web Analytics](https://www.cloudflare.com/en-gb/web-analytics/) on my Docusaurus site.

To enable this I need to add a script to every page. Note the `data-cf-beacon` value is a JSON object.

```
<script defer src='https://static.cloudflareinsights.com/beacon.min.js' data-cf-beacon='{"token": "1234"}'></script>
```

In `docusaurus.config.ts` I added the Cloudflare script to `scripts`:

```typescript
const config: Config = {
  scripts: [
    {
      // Cloudflare Web Analytics
      defer: true,
      async: false,
      src: "https://static.cloudflareinsights.com/beacon.min.js",
      "data-cf-beacon": "{\"token\": \"1234\"}",
    },
  ]
}
```

However, Docusaurus escapes the quotes in the JSON value, meaning that the JSON is invalid.

```html
<!doctype html>
<html lang="en" dir="ltr" class="plugin-pages plugin-id-default" data-has-hydrated="false">
<head>
  <meta charset="UTF-8">
  <meta name="generator" content="Docusaurus v3.3.2">

  [...]

  <script defer="defer" src="https://static.cloudflareinsights.com/beacon.min.js"
          data-cf-beacon="{&quot;token&quot;:&quot;1234&quot;}"></script>
```

### Reproducible demo

https://stackblitz.com/edit/github-sqjruf?file=docusaurus.config.ts

### Steps to reproduce

See above

### Expected behavior

Either:

1. The quotes are not escaped, and the literal value is rendered.
2. There is some way to disable escaping.

   I see that [Eta has a function for modifying escaping](https://eta.js.org/docs/api/configuration), perhaps it could be used?

   ```javascript
   /** Function to XML-sanitize interpolations */
   escapeFunction: (str: unknown) => string
   ```

I would also like it if the solution was documented.

### Actual behavior

Docusaurus escapes the quotes in the JSON value, meaning that the JSON is invalid.

It's not easy to figure out a workaround. I think manually updating the [`ssrTemplate`](https://docusaurus.io/docs/api/docusaurus-config#ssrTemplate) works:

```javascript
// docusaurus.config.ts

import defaultSSRTemplate from '@docusaurus/core/lib/templates/ssr.html.template';

const config: Config = {

  ssrTemplate: defaultSSRTemplate
  .replace(
      "</head>",
      "<script defer src='https://static.cloudflareinsights.com/beacon.min.js' data-cf-beacon='{\"token\": \"1234\"}'></script>\n</head>"
  ),

```

### Your environment

- Public source code: https://github.com/adamko-dev/dokkatoo/blob/ba9fe26aaa0add74e78f7527282f5e46fd24999e/modules/docs/site/docusaurus.config.ts#L36-L43
- Public site URL: https://adamko-dev.github.io/dokkatoo/
- Docusaurus version used: 3.3.2
- Environment name and version (e.g. Chrome 89, Node.js 16.4): n/a
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): n/a

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by dfinster** [comment]
I'm also running into this broken JSON object when using Cloudflare Analytics on Docusaurus 3.3.2. Unfortunately, I can't share our source code until the site is made public later this year.

It seems strange that I only find one other person (@aSemy) with this issue when I [search for instances](https://github.com/search?q=data-cf-beacon+path%3A%2F%28%5E%7C%5C%2F%29docusaurus%5C.config%5C.ts%24%2F&type=code) in `docusaurus.config.ts`. It seems like it would be more common than that. However, I found a few more examples when [expanding the search](https://github.com/search?q=data-cf-beacon+path%3A%2F%28%5E%7C%5C%2F%29docusaurus%5C.config%5C.js%24%2F&type=code) from Typescript to Javascript. Some people [ignored the issue](https://github.com/search?q=data-cf-beacon%27%3A+%27%7B%5C%22token%5C%22%3A+path%3A%2F%28%5E%7C%5C%2F%29docusaurus%5C.config%5C.js%24%2F&type=code) and published sites with broken JSON objects. Maybe Cloudflare Analytics isn't as popular as I thought, or maybe people are using [automatic mode](https://developers.cloudflare.com/analytics/web-analytics/getting-started/#sites-proxied-through-cloudflare), which can directly inject the beacon for sites proxied through Cloudflare.

### Other attempts

I see other people trying various approaches to fix this JSON quoting issue:

- [Escape with backticks](https://github.com/learntocloud/learn-to-cloud/blob/145ee987e4edfe040d3b78a7b071d975719f037e/docusaurus.config.js#L33).
  - It doesn't work, view source line 11 [of the site](https://learntocloud.guide/). 
- [HTML-encode the double quotes](https://github.com/sgraband/jsonforms2-website/blob/4f0304b0b9f0f74681a21f0865b152b01f2ba540/docusaurus.config.js#L18).
  - It doesn't work, view source line 10 [of the site](https://jsonforms.io/).
- [Inject `postBodyTags`](https://github.com/espanso/website/blob/4b015ba3073d30b76273e6218447921309000289/docusaurus.config.js#L145) into the ([soon to be deprecated](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-google-analytics)) Google Analytics plugin, while escaping the double quotes.
  - _**This does work**_, view source line 16 [of the site](https://espanso.org/). It doesn't seem like the best approach.

### Things I tried

I tried escaping with backticks (point 1 above) because it was a quick test. I can confirm it doesn't work. I also tried swizzling (wrap mode) the footer component to insert the link:

```typescript
// theme/footer/index.tsx

import React from 'react';
import Footer from '@theme-original/Footer';
import type FooterType from '@theme/Footer';
import type { WrapperProps } from '@docusaurus/types';

type Props = WrapperProps<typeof FooterType>;

export default function FooterWrapper(props: Props): JSX.Element {
  return (
    <>
      <Footer {...props} />
      <script
        defer
        src='https://static.cloudflareinsights.com/beacon.min.js'
        data-cf-beacon='{"token": "00000000000000000redacted0000000"}'
      ></script>
    </>
  );
}
```

This swizzle attempt didn't work.

### My conclusion

At this point, I think the "least worst" option for hosting the beacon script in Docusaurus is @aSemy's [workaround](https://github.com/adamko-dev/dokkatoo/commit/99508fdc5e4b343734300fd37dc23cd927e8d258). I'll keep monitoring this issue and hope there's a cleaner solution someday.

Other options might be:

- Create a new [behavior plugin](https://docusaurus.io/docs/api/plugins#behavior-plugins) for Cloudflare Analytics, but I don't know if it's worth the extra effort for a uni-tasking solution.
- @aSemy's suggestion to explore [Eta's `escapeFunction`](https://eta.js.org/docs/api/configuration), which could be useful for other things, if I'm understanding that possibility correctly.

[Cloudflare's automatic injection](https://developers.cloudflare.com/analytics/web-analytics/getting-started/#sites-proxied-through-cloudflare) might be how most people do it. 🤷🏻

**Comment by slorber** [comment]
It is possible there is a bug in the way we escape HTML attribute values. It is quite old historical code that's not really related to ETA https://github.com/facebook/docusaurus/blob/61f71f6b69a36350dc0c28fb291dfc591904ee27/packages/docusaurus/src/server/htmlTags.ts#L69

However, are we even sure the output of Docusaurus is wrong.

```html

<!doctype html>
<html lang="en" dir="ltr" class="plugin-pages plugin-id-default" data-has-hydrated="false">
<head>
  <meta charset="UTF-8">
  <meta name="generator" content="Docusaurus v3.3.2">
  <script id="testscript" defer="defer" src="https://static.cloudflareinsights.com/beacon.min.js" data-cf-beacon="{&quot;token&quot;:&quot;1234&quot;}"></script>
<head>
<body>body</body>
</html>
```

For browsers, the following HTML looks valid and gets decoded properly to a JSON value.

![CleanShot 2024-05-30 at 12 27 36](https://github.com/facebook/docusaurus/assets/749374/75e0458d-111c-450a-9b98-a554ed245d56)

![CleanShot 2024-05-30 at 12 28 11](https://github.com/facebook/docusaurus/assets/749374/1559a134-8255-4292-ae80-53338eeb2934)


Can you prove that the output of Docusaurus is wrong, and that it's not a bug in the script you load?

Is there anything written in the HTML spec mentioning that `&quot;` encoding is invalid as an attribute?

**Comment by aSemy** [comment]
> For browsers, the following HTML looks valid and gets decoded properly to a JSON value.
> 
> ![CleanShot 2024-05-30 at 12 27 36](https://private-user-images.githubusercontent.com/749374/335143570-75e0458d-111c-450a-9b98-a554ed245d56.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNjY1NjQsIm5iZiI6MTcxNzA2NjI2NCwicGF0aCI6Ii83NDkzNzQvMzM1MTQzNTcwLTc1ZTA0NThkLTExMWMtNDUwYS05Yjk4LWE1NTRlZDI0NWQ1Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUzMFQxMDUxMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04Yzc0MmI4OWRiZDU4MjViZjAzY2NkZThjZWVkMDlmNzg3MTQ3MWZkNzg4OWMwYzBmODgyYTc5ZWM0ZDVjYWY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.NlO99OPJ43Ifyo0N2XD4Dz1OrLvPFre4fQCTZfaZTM4)

True, it looks correct, but I think that is Chrome trying to be friendly. When I right click and edit the element, I see that the value is not valid JSON.

<img width="996" alt="image" src="https://github.com/facebook/docusaurus/assets/897017/77e22be3-6c0f-4d83-9596-1516afb9b687">




> Can you prove that the output of Docusaurus is wrong, and that it's not a bug in the script you load?

I think I'm confused about what you're asking, but I'll try and answer. The attribute value that I configured is valid JSON `{"token": "1234"}`. However, Docusaurus produces invalid JSON `{&quot;token&quot;: &quot;1234&quot;}`.

<img width="688" alt="image" src="https://github.com/facebook/docusaurus/assets/897017/4206baa5-9915-4ecc-ac02-fed968c65762">

This problem occurs even before the external script is loaded, so it's not a bug in the script. 

> 
> Is there anything written in the HTML spec mentioning that `&quot;` encoding is invalid as an attribute?

What?

> Please show me the problem outside of Docusaurus:
> 
> * show me a manually crafted HTML file with the bad encoding
  
  ```
  <script defer src="https://static.cloudflareinsights.com/beacon.min.js" data-cf-beacon="{&quot;token&quot;: &quot;1234&quot;}"></script>
  ```

> * show me a manually crafted HTML file with the encoding you expect
  
  ```
  <script defer src="https://static.cloudflareinsights.com/beacon.min.js" data-cf-beacon='{"token": "1234"}'></script>
  ```

> * show me how you figure out the encoding is bad, do you see any error?


When the attribute quotes are escaped as `&quot;`, I get zero hits in Cloudflare Web Analytics. [After I applied a workaround](https://github.com/adamko-dev/dokkatoo/commit/99508fdc5e4b343734300fd37dc23cd927e8d258), I see hits in Cloudflare.

> What produces that error? Show me the JS script part reading the data attribute for example

There is no error. 

> TLDR I need a minimal repro, probably not even involving Docusaurus


I can reproduce the issue as follows:

1. In `docusaurus.config.ts` I added the Cloudflare script to `scripts`:
    
    ```
    const config: Config = {
      scripts: [
        {
          // Cloudflare Web Analytics
          defer: true,
          async: false,
          src: "https://static.cloudflareinsights.com/beacon.min.js",
          "data-cf-beacon": "{\"token\": \"1234\"}",
        },
      ]
    }
    ```
2. However, Docusaurus escapes the quotes in the JSON value, meaning that the JSON is invalid.
    
    ```html
    <!doctype html>
    <html lang="en" dir="ltr" class="plugin-pages plugin-id-default" data-has-hydrated="false">
    <head>
      <meta charset="UTF-8">
      <meta name="generator" content="Docusaurus v3.3.2">
    
      [...]
    
      <script defer="defer" src="https://static.cloudflareinsights.com/beacon.min.js"
              data-cf-beacon="{&quot;token&quot;:&quot;1234&quot;}"></script>
    ```


I don't understand what you mean about trying not to reproduce the problem without Docusaurus.
If I don't use Docusaurus then I don't use `docusaurus.config.ts`, and so I can add `<script defer src="https://static.cloudflareinsights.com/beacon.min.js" data-cf-beacon='{"token": "1234"}'></script>` manually. But that's not a reproducer for the issue...

**Comment by slorber** [comment]
What I mean is: how can you be sure Docusaurus is the problem, and not Cloudflare Web Analytics?

I believe what we do is fine, and that the bug might be somewhere else. 

Yes, I'm asking you to troubleshoot the script Cloudflare provides you.

https://gist.github.com/slorber/bec62fcfb278d5a569e2d2c9c716504a

![CleanShot 2024-05-30 at 15 46 59](https://github.com/facebook/docusaurus/assets/749374/69c1bb76-4441-4d32-90d8-28d03acda630)

![CleanShot 2024-05-30 at 15 49 34](https://github.com/facebook/docusaurus/assets/749374/fc72f4e9-53df-4007-aa95-fd61647414ca)

There might be other places in that script that break, but the code above found in the script seems to work fine.

If something is broken, please tell me what exactly by isolating things. You shouldn't need to depend on Docusaurus nor a third-party cloudflare script to prove your point, and a simple HTML file should be enough to reproduce the problem, if there is a problem in the first place.

**Comment by aSemy** [comment]
> What I mean is: how can you be sure Docusaurus is the problem, and not Cloudflare Web Analytics?

Docusaurus is the problem because it is escaping a value, meaning that Docusaurus converts valid JSON into invalid JSON. Cloudflare Web Analytics is not involved with how Docusaurus generates the HTML.

> If something is broken, please tell me what exactly by isolating things.

I am trying to provide this details, but I'm really really not sure what is not clear about. If you need more details, then I would be happy to provide them, but I'm really not sure how else I can explain it. 

1. I configure Docusaurus to add a `<script>` tag.
2. I need the script tag to contain a specific attribute with a specific value.
3. Docusaurus escapes the attribute value, meaning that the attribute no longer contains the correct value.
4. I expect that Docusaurus either doesn't escape the value, or provides a way of preventing the escaping.

I happen to need this value for analytics, but that's just an example usecase, and I think it's confusing the situation. The analytics script isn't generating the HTML, that's done by Docusaurus.

> You shouldn't need to depend on Docusaurus nor a third-party cloudflare script to prove your point, and a simple HTML file should be enough to reproduce the problem, if there is a problem in the first place.

How can I demonstrate that Docusauarus converts `"foo"` into `&quot;foo&quot;` without using Docusauarus? 

> If the encoding we use is fine and spec-compliant, then it is unfair to ask us to change it to work around a bug found in another system.
> 
> If it is proven that our encoding is wrong, then we'll make a change. And having 0 analytics on Cloudflare does not prove that the encoding is wrong.

**Comment by dfinster** [comment]
I'm sad that we've lost the comment history here, because it's a good opportunity for me to learn new things.

Namely:

1. A technique [I didn't know about](https://github.com/adamko-dev/dokkatoo/blob/99508fdc5e4b343734300fd37dc23cd927e8d258/modules/docs/site/docusaurus.config.ts#L40) for injecting code with `defaultSSRTemplate
  .replace()`
2. The _**possibility**_ that an HTML-encoded JSON object can be correctly decoded with `JSON.parse()`. We are talking about the difference between these two strings appearing in the rendered HTML:
  - `data-cf-beacon='{"token": "1234"}'`
  - `data-cf-beacon="{&quot;token&quot;: &quot;1234&quot;}"`
 
I made the _**assumption**_ that the second example in (2) would break the script, but I didn't thoroughly test it. @slorber suggests that it's valid. When I have time I'll make a minimum viable example to test this out. If true, then I can strike the word "possibility" above and I will have learned a new thing. That's exciting to me.

All I really know at this point is that `scripts: []` in `docusaurus.config.ts` HTML-encodes the double-quotes, as do JSX/TSX components.  I think @slorber is correct to ask for a reproducible example, but it looks like (from what I remember of the now-deleted comments) that there was a misunderstanding about what was being requested and why. 

Peace to you all.

**Comment by slorber** [comment]
> The possibility that an HTML-encoded JSON object can be correctly decoded with JSON.parse().

It's not really HTML-encoded JSON, it's just JSON that is transferred as part of an HTML payload. It's not `JSON.parse()` that decodes it, once it's in the DOM it's already decoded.
